ArmVirtPkg/ArmVirtKvmTool.dsc

```diff
@@ -94,11 +94,6 @@
 
 [BuildOptions]
   *_*_*_CC_FLAGS = -D DISABLE_NEW_DEPRECATED_INTERFACES
-  #
-  # We need to avoid jump tables in SEC and BASE modules, so that the PE/COFF
-  # self-relocation code itself is guaranteed to be position independent.
-  #
-  GCC:*_*_*_CC_XIPFLAGS = -fno-jump-tables

```

Comment:
Why can we just drop this?

Reply:
[-fno-jump-tables](https://gcc.gnu.org/onlinedocs/gcc/Code-Gen-Options.html)
Do not use jump tables for switch statements even where it would be more efficient than other code generation strategies. This option is of use in conjunction with -fpic or -fPIC for building code that forms part of a dynamic linker and cannot reference the address of a jump table. On some targets, jump tables do not require a GOT and this option is not needed. 

All images generated by ImageTool are position dependent [(-fno-pic -fno-pie)](https://github.com/acidanthera/audk/blob/master/BaseTools/Conf/tools_def.template#L1880) 

I [removed](https://github.com/acidanthera/audk/pull/27/commits/cc272bb838b65735b173bb54581245af2630f415) the use case of that flag entirely and am still  able to boot Debian 11 for aarch64/arm.

Reply:
Position-dependent codegen actually makes the problem worse, as RIP-based addressing is less likely to be used. The description of the flag is ambiguous - it does not explicitly state the option is without effect for -fno-pic, only that it makes sense to be used with -fPIC to avoid absolute addressing (that's why the last sentence mentions GOT). While the binaries as a whole may have relocs for absolute addresses, the idea (and requirement!) of self-relocation is that the early code performing the relocation is in fact position-independent without relocs.

Something that relies on this and in fact is not -fno-pic -fno-pie is StandaloneMmCore [1]. StandaloneMmCoreEntryPointLib self-relocates the image very soon after start [2]. I'm not sure whether it works as-is as I never got to test ARM targets before. Booting one OS on one target without verifying all platforms with self-relocating modules is not a sufficient test at all to remove such functionality.

StandaloneMmCore is not affected by this change, as it was specific to SEC and PEI modules, but I do not have a good overview over all ARM platform trees to tell whether really all of them are safe to build now with this removed. StandaloneMmCore just is the one example I'm somewhat familiar with how it works.

[1] https://github.com/acidanthera/audk/blob/b877d75f50b3a01b09f62ee3265adfa5c6da9622/StandaloneMmPkg/Core/StandaloneMmCore.inf#L80-L87

[2] https://github.com/acidanthera/audk/blob/b877d75f50b3a01b09f62ee3265adfa5c6da9622/StandaloneMmPkg/Library/StandaloneMmCoreEntryPoint/Arm/StandaloneMmCoreEntryPoint.c#L351-L400

Reply:
This is an interesting and fairly dangerous example. The comment in the .dsc is very misleading.

@mhaeuser, does it really mean that SEC/PEI code on ARM may be moved without updating the relocations? To me this violates the mere basics of image loading…


If there is such a requirement, I request:

- The flag is to be kept.
- The comment changed to explain that no relocations are allowed in SEC/PEI on ARM due to moving code. @mhaeuser, I would welcome if you suggest more clear wording.
- No relocations must be enforced by ImageTool. I am not exactly sure how to do this, but this should be done.

Reply:
@vit9696 No, I think there's a slight misunderstanding. No relocations are allowed *in the self-relocation code*. As you're aware, I'm not terribly familiar with ARM, but I was told by Ard or Leif (don't remember) that for StandaloneMmCore and other "initial" modules (which SEC would be, thus the flag being SEC/PEI only makes sense to me), they may indeed have load addresses unknown at load-time (which is not a concern with x86 SEC, obviously) - whether that depends on the exact hardware platform, I don't know. The idea, as you can see in [2], is that the module relocates itself. This process, from the entry point till `UefiImageRelocateImageInplaceForExecution()` returns, must work position-independently without relocs. As after it returned relocs have been processed, code that relies on relocs is safe from that point onwards.

As you can see in [1], it passes `-Wl,-z,text` to the linker, so it will warn about any relocs to .text and other read-only segments. There may of course in theory be some dependency on .data (et al.) relocs, but our loader does not use global variables, and I suppose that is the "convention" to make this safe. I don't think ImageTool needs to be involved with how things work right now, at least for StandaloneMmCore.